src/installer/managed/Microsoft.NET.HostModel/ResourceUpdater.cs

```diff
@@ -326,22 +124,65 @@ public ResourceUpdater AddResource(byte[] data, IntPtr lpType, IntPtr lpName)
         /// </summary>
         public ResourceUpdater AddResource(byte[] data, string lpType, IntPtr lpName)
         {
-            if (hUpdate.IsInvalid)
+            if (!IsIntResource(lpName))
             {
-                ThrowExceptionForInvalidUpdate();
+                throw new ArgumentException("AddResource can only be used with integer resource names");
             }
+            if (_resourceData == null)
+                throw ThrowExceptionForInvalidUpdate();
 
-            if (!IsIntResource(lpName))
+            _resourceData.AddResource((ushort)lpName, lpType, LangID_LangNeutral_SublangNeutral, data);
+
+            return this;
+        }
+
+        // see https://learn.microsoft.com/windows/win32/debug/pe-format
+        private static class Offsets
+        {
+            private const int PESignatureSize = sizeof(int);
+            private const int CoffHeaderSize = 20;
+            public const int PEHeaderSize = PESignatureSize + CoffHeaderSize;
+            public const int OneSectionHeaderSize = 40;
+            public const int DataDirectoryEntrySize = 8;
+
```

Comment:
I wonder if we could somehow share this with https://github.com/dotnet/runtime/blob/main/src/installer/managed/Microsoft.NET.HostModel/AppHost/PEUtils.cs.
Definitely not for this PR - but if you had some ideas, it would be really nice.

Reply:
I moved `Offsets` to `PEOffsets` internal class and share offsets between `ResourceUpdater` and `PEUtils` in 6e8120b.